libraries/ESP8266WebServer/examples/ServerSentEvents/ServerSentEvents.ino

```diff
@@ -0,0 +1,218 @@
+
+ESP8266WebServer server(port);
+
+#define SSE_MAX_CHANNELS 8  // in this simplified example, only eight SSE clients subscription allowed
+struct SSESubscription {
+  uint32_t clientIP;
+  WiFiClient client;
+  Ticker keepAliveTimer;
+} subscription[SSE_MAX_CHANNELS];
+uint8_t subscriptionCount = 0;
+
+unsigned short sensorA = 0, sensorB = 0;   //Simulate two sensors
+Ticker update;
+
+void handleNotFound() {
+  Serial.println(F("Handle not found"));
+  String message = "Handle Not Found\n\n";
+  message += "URI: ";
+  message += server.uri();
+  message += "\nMethod: ";
+  message += (server.method() == HTTP_GET) ? "GET" : "POST";
+  message += "\nArguments: ";
+  message += server.args();
+  message += "\n";
+  for (uint8_t i = 0; i < server.args(); i++) {
+    message += " " + server.argName(i) + ": " + server.arg(i) + "\n";
+  }
+  server.send(404, "text/plain", message);
+}
+
+void SSEKeepAlive() {
+  for (uint8_t i = 0; i < SSE_MAX_CHANNELS; i++) {
+    if (!(subscription[i].clientIP)) {
+      continue;
+    }
+    WiFiClient client = subscription[i].client;
+    if (client.connected()) {
+      Serial.printf_P(PSTR("SSEKeepAlive - client is still listening on channel %d\n"), i);
+      client.println(F("event: event\ndata: { \"TYPE\":\"KEEP-ALIVE\" }\n"));   // Extra newline required by SSE standard
+    } else {
+      Serial.printf_P(PSTR("SSEKeepAlive - client not listening on channel %d, remove subscription\n"), i);
+      subscription[i].keepAliveTimer.detach();
+      client.flush();
+      client.stop();
+      subscription[i].clientIP = 0;
+      subscriptionCount--;
+    }
+  }
+}
+
+// SSEHandler handles the client connection to the event bus (client event listener)
+// every 60 seconds it sends a keep alive event via Ticker
+void SSEHandler(uint8_t channel) {
+  WiFiClient client = server.client();
+  SSESubscription &s = subscription[channel];
+  if (s.clientIP != uint32_t(client.remoteIP())) { // IP addresses don't match, reject this client
+    Serial.printf_P(PSTR("SSEHandler - unregistered client with IP %s tries to listen\n"), server.client().remoteIP().toString().c_str());
+    return handleNotFound();
+  }
+  client.setNoDelay(true);
+  client.setSync(true);
+  Serial.printf_P(PSTR("SSEHandler - registered client with IP %s is listening\n"), IPAddress(s.clientIP).toString().c_str());
+  s.client = client; // capture SSE server client connection
+  server.setContentLength(CONTENT_LENGTH_UNKNOWN); // the payload can go on forever
+  server.sendContent_P(PSTR("HTTP/1.1 200 OK\nContent-Type: text/event-stream;\nConnection: keep-alive\nCache-Control: no-cache\nAccess-Control-Allow-Origin: *\n\n"));
+  s.keepAliveTimer.attach_scheduled(30.0, SSEKeepAlive);  // Refresh time every 30s for demo
+}
+
+void handleAll() {
+  const char *uri = server.uri().c_str();
+  if (strncmp_P(uri, PSTR("/rest/events/"), sizeof("/rest/events"))) {
+    return handleNotFound();
+  }
+  uri += sizeof("/rest/events");
```

Comment:
isn't there a missing / here? I'm guessing the uri should be "rest/events/<channel>"

Reply:
No, thanks to C strings having the \0.  `sizeof("abc") == 4` since there's the terminal NUL.  This uses the \0 to skip the '/'.  The script does seem to work well in my own local testing.